### PR TITLE
Optimize transformer loops

### PR DESCRIPTION
**Pull request by mikowals**
Marked as draft because I built off of the PR #38 branch.

This gets about a 25% speed up by using Mojo's `parallelize` and `vectorize` on the loops in the transformer forward pass.  I also removed the division of `num_cores()` by two but the speed up was apples to apples comparisons of both versions running with all cores.  This was on an 8 core Codespace instance.

Looking at the last commit (to ignore the shift to Tensors) you can see it is just applying the standard for-loop to parallelize (or vectorise) conversion in a some key places.  The main impact is making the loops over attention heads parallel.  Those loops have other loops nested within.  So I believe the biggest impact is from changing 4 lines of code.  So whether the shift to Tensors is adopted or not adding these optimisations is probably worth doing.

### CODE REVIEW

**Comment by tairov** [comment]
25% speedup is really impressive, I was considering the same speedup , since `llama2.c` also contain `pragma omp` on the same loop -- https://github.com/karpathy/llama2.c/blob/master/run.c#L290
But then my profilings over python code was showing that most of the time is consuming in matmul. 

Seems that this forward pass is another hot-code section.

**Comment by tairov** [comment]
I was able to reproduce the speedup, but I had to set Runtime threads = `num_cores() // 2` for my target host

**self.rt = Runtime(num_cores())**
```
mojo llama2.mojo stories15M.bin -t 0.0
...
achieved tok/s:  203.32717190388169
```

**self.rt = Runtime(num_cores() // 2)**
```
...
achieved tok/s:  440.88176352705409
```

`440 tok/s` -- is really cool
for comparison `llama2.c` performance in OMP mode on target machine : `435 tok/s`

**Comment by tairov** [comment]
Only 3 cores/vcores (out of 6) are busy 
image

tinyllama performance :
```
achieved tok/s:  6.84
```

File: llama2.mojo [hunk 1]
```diff
@@ -24,113 +25,75 @@ alias PointerString = Pointer[UInt8]
...
+
+    fn __init__(inout self, t: TensorF32, layer: Int):
+        let elements_per_layer = t.num_elements() // t.dim(0)
+        self._data = t.data().offset(layer * elements_per_layer)
+        if t.rank() == 2:
+            self._shape = TensorShape(t.dim(1))
+        elif t.rank() == 3:
+            self._shape = TensorShape(t.dim(1), t.dim(2))
+        else:
+            self._shape = TensorShape(t.dim(1), t.dim(2), t.dim(3))
```
[the remark this answers is not in the document]
Comment (mikowals) [reply llama2.mojo]: [Addressed]([24b9e1e](https://github.com/tairov/llama2.mojo/pull/39/commits/24b9e1e49f38d85b42176f652f7d52937c747a45) by raising errors of for unexpected rank.

The compiler also required me to set _shape so I have done this and added a comment explaining that.

File: llama2.mojo [hunk 2]
```diff
@@ -24,113 +25,75 @@ alias PointerString = Pointer[UInt8]
...
+    fn __init__(inout self, t: TensorF32, layer: Int, row: Int):
+        let elements_per_layer = t.num_elements() // t.dim(0)
+        let elements_per_row = elements_per_layer // t.dim(1)
+        self._data = t.data().offset(
+            layer * elements_per_layer + row * elements_per_row
+        )
+        if t.rank() == 3:
+            self._shape = TensorShape(t.dim(2))
+        else:
+            self._shape = TensorShape(t.dim(2), t.dim(3))
```
[the remark this answers is not in the document]
Comment (mikowals) [reply llama2.mojo]: [Addressed]([24b9e1e](https://github.com/tairov/llama2.mojo/pull/39/commits/24b9e1e49f38d85b42176f652f7d52937c747a45) by raising errors of for unexpected rank.

The compiler also required me to set _shape so I have done this and added a comment explaining that.

File: llama2.mojo [hunk 3]
```diff
@@ -630,143 +635,141 @@ fn transformer(
...
+        rmsnorm(state.xb, state.x, TensorSlice(weights.rms_ffn_weight, l))
 
         # Calculate self.w1(x) and self.w3(x) for FFN
-        tmpw.set_buf_ptr(weights.w1.data.offset(l * dim * hidden_dim), hidden_dim, dim)
-        matmul(state.hb, state.xb, tmpw, state.rt)
-
-        tmpw.set_buf_ptr(weights.w3.data.offset(l * dim * hidden_dim), hidden_dim, dim)
-        matmul(state.hb2, state.xb, tmpw, state.rt)
-
-        # Apply SiLU activation function (silu(x) = x * sigmoid(x))
```
[the remark this answers is not in the document]
Comment (mikowals) [reply llama2.mojo:733]: I joined the two for-loops and moved the comments inside the loop above the lines they explain.  Leaving the comments above the for loop seemed lees clear now that the loop did two things.

File: llama2.mojo (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikowals) [reply llama2.mojo]: now really fixed by [[028b12d](https://github.com/tairov/llama2.mojo/pull/39/commits/028b12d37abd2aed49e4c44428ff8f4db2ce5a28).

File: llama2.mojo [hunk 4]
```diff
@@ -24,113 +25,75 @@ alias PointerString = Pointer[UInt8]
...
+    fn simd_load[
+        nelts: Int
+    ](self, indices: StaticIntTuple[2]) -> SIMD[DType.float32, nelts]:
+        return self._data.simd_load[nelts](indices[0] * self._shape[1] + indices[1])
 
-    @always_inline
-    fn load[nelts: Int](self, z: Int, y: Int, x: Int) -> SIMD[DType.float32, nelts]:
-        return self.data.simd_load[nelts](z * self.layers + y * self.cols + x)
+    fn __getitem__(self, idx: Int) -> SIMD[DType.float32, 1]:
+        return self._data.simd_load[1](idx)
```
[the remark this answers is not in the document]
Comment (mikowals) [reply llama2.mojo:101]: yes, consistency is better.  Thanks.  Changed in [28664b8](https://github.com/tairov/llama2.mojo/pull/39/commits/28664b87f5413725f1de091d065436f6aae7fd09).

**Comment by mikowals** [comment]
We can progress to merge this PR only.  I think the commit history should be squashed before merge but I will keep "replace Matrix with Tensor" and "vectorize and parallelize loops in transformer" as separate commits since they are unrelated changes.

**Comment by tairov** [comment]
@mikowals well, if you prefer the keep 2 commits only it would be better. Otherwise GH has a cool feature  :)

image

**Comment by Saddibek1** [comment]
Have you tested num_cores//1 ?)

**Pull request merged**